src/LockTable.cc

```diff
@@ -0,0 +1,225 @@
+
+/**
+ * Constructor for LockTable.
+ *
+ * \param numBuckets
+ *      Locks in a LockTable are stored in one or more buckets.  This parameter
+ *      sets the number of buckets that should be used for this new LockTable.
+ *      This value should be a power of two.
+ *
+ * \throw Exception
+ *      An exception is thrown if the resulting number of buckets is 0.
+ */
+LockTable::LockTable(uint64_t numBuckets)
+    : bucketIndexHashMask(BitOps::powerOfTwoLessOrEqual(numBuckets) - 1)
+    , buckets()
+{
+    if (numBuckets != (bucketIndexHashMask + 1)) {
+        RAMCLOUD_LOG(DEBUG,
+                     "LockTable truncated to %lu buckets "
+                     "(nearest power of two)",
+                     (bucketIndexHashMask + 1));
+    }
+
+    if ((bucketIndexHashMask + 1) == 0)
+        throw Exception(HERE, "LockTable numBuckets == 0?!");
+
+    void *buf  = Memory::xmemalign(
+            HERE,
+            sizeof(CacheLine),
+            (bucketIndexHashMask + 1) * sizeof(CacheLine));
+    memset(buf, 0, (bucketIndexHashMask + 1) * sizeof(CacheLine));
+    buckets = static_cast<CacheLine *>(buf);
+}
+
+/**
+ * Destructor for LockTable.
+ */
+LockTable::~LockTable() {
+    // Free any extra xmalloced CacheLines.
+    for (uint64_t i = 0; i < bucketIndexHashMask + 1; i++) {
+        CacheLine* cacheLine = buckets[i].next;
+        while (cacheLine != NULL) {
+            CacheLine* next = cacheLine->next;
+            free(cacheLine);
+            cacheLine = next;
+        }
+    }
+    free(buckets);
+}
+
+/**
+ * Blocks until the lock with the provided lockId can be acquired.
+ *
+ * \param lockId
+ *      ID of lock to be acquired.  ID must not be zero.
+ */
+void
+LockTable::acquireLock(uint64_t lockId)
+{
+    while (!tryAcquireLock(lockId))
+        continue;
+}
+
+/**
+ * Check if lock with the provided lockId is currently acquired.
+ *
+ * \param lockId
+ *      ID of lock to be acquired.  ID must not be zero.
+ * \return
+ *      TRUE if the lock is currently acquired, FALSE otherwise.
+ */
+bool
+LockTable::isLockAcquired(uint64_t lockId)
+{
+    // Find the right bucket.
+    uint64_t bucketIndex = (lockId & bucketIndexHashMask);
+    CacheLine* cacheLine = &buckets[bucketIndex];
+
+    // Lock the bucket
+    BucketLock* bucketLock =
+            reinterpret_cast<BucketLock*>(&cacheLine->entries[0]);
+    std::lock_guard<BucketLock> lock(*bucketLock);
+
+    // The zeroth entry in the first CacheLine is the BucketLock so start the
+    // index at 1.
+    uint32_t entryIndex = 1;
+    while (true) {
+        for (; entryIndex < ENTRIES_PER_CACHE_LINE; entryIndex++) {
+            if (cacheLine->entries[entryIndex] == lockId) {
+                return true;
+            }
+        }
+        if (cacheLine->next != NULL) {
+            entryIndex = 0;
+            cacheLine = cacheLine->next;
+        } else {
+            break;
+        }
+    }
+    return false;
+}
+
+/**
+ * Releases the lock with the provided lockId.
```

Comment:
Mention that if there is no lock with that lockId, then the method does nothing.